### DISCUSSION TITLE
Calculate Distance in YOLOv8

### DISCUSSION THREAD

**Discussion by CrispyHands** [ultralytics/ultralytics#5779] [Q&A]
I have predicted with yolov8 using custom dataset. but, I still don't understand how to get the bounding box and then calculate the way between the bounding boxes using euclidean distance?

**Comment by ahillsavio1** [comment]
I can help you if you meant "getting the coordinates of the bounding box"

```
def yolo_predict(img):
    model = YOLO(r'path\to\model.pt')

    results = model.predict(
        source=img,
        imgsz=640,
        ..whatever parameters you need..
    )
    # This returns the coordinates of the bounding box, specifically top left and bottom right corners.
    for result in results:
        coords = result.boxes.xyxy  # There are methods other than .xyxy too, just search it up in the internet.

    return coords.numpy()
```
I hope this helps you.

**Reply by CrispyHands** [reply]
![WhatsApp Image 2023-10-19 at 17 11 45_1c493e6a](https://github.com/ultralytics/ultralytics/assets/37016899/69baf70b-a3ff-4bc2-8537-af18a8c03391)

i mean look like this.

**Reply by ahillsavio1** [reply]
Im not sure if there is a method available for this (I am a newbie myself) but I guess you can manually write a code for this. Like, you can use the coordinates you fetched and put them in the Euclidean's distance formula as needed.

**Reply by CrispyHands** [reply]
i'm have a code already just need to change that coordinate too look like that?
do you have a discord? probably i can share the code there.

**Reply by ahillsavio1** [reply]
Yeah I do, do tell me your id, ill connect with you.

**Reply by pderrenger** [reply]
@ahillsavio1 i'm glad to hear that you have a code ready for calculating distances. However, I must clarify that as the maintainer of the Ultralytics YOLOv8 repo, I can only provide support through official channels like GitHub issues and our documentation.

To calculate the Euclidean distance between bounding boxes, you can extract the coordinates from the `results` object after prediction. Here's a concise example of how you might do that:

```python
from ultralytics import YOLO
import numpy as np

# Load your model
model = YOLO('path/to/your/model.pt')

# Perform prediction
results = model.predict(source='path/to/your/image.jpg')

# Extract bounding box coordinates
boxes = results.xyxy[0]  # xyxy format is [x1, y1, x2, y2]

# Calculate Euclidean distance between two bounding boxes (e.g., box1 and box2)
def euclidean_distance(box1, box2):
    center1 = np.array([(box1[0] + box1[2]) / 2, (box1[1] + box1[3]) / 2])
    center2 = np.array([(box2[0] + box2[2]) / 2, (box2[1] + box2[3]) / 2])
    distance = np.linalg.norm(center1 - center2)
    return distance

# Example usage with the first two detected bounding boxes
if len(boxes) > 1:
    distance = euclidean_distance(boxes[0], boxes[1])
    print(f"Distance between boxes: {distance}")
```

Please replace `'path/to/your/model.pt'` and `'path/to/your/image.jpg'` with the actual paths to your model and image.

Remember to check the documentation for any updates or additional features that might help with your task. If you have further questions or issues, feel free to open an issue in the GitHub repository, and we'll be happy to assist you there. 🚀

**Reply by SMRaza1009** [reply]
Dear,

I also have a query. I have traffic light custom dataset. I want to calculate the distance of traffic light signal from a car video. I am currently trying to do with YOLOv8 distance calculator but it seems failed in my case. Because it calculates the distance between the object. However, I need to calculate distance of traffic light from a moving car. 

Can you assist me this regard? I shall be thankful to you!

**Reply by pderrenger** [reply]
Hello!

Absolutely, happy to help! 🚗💡 For calculating the distance of a traffic light from a moving car, you might need to integrate velocity or frame-by-frame displacement if you're working directly with video data.

If you're using bounding box sizes as a proxy for distance (where smaller boxes represent farther away objects), you could analyze the size change over frames for a rough distance estimation.

Here's a quick snippet on how to get the bounding box of your traffic light and potentially use its size for distance estimation:

```python
from ultralytics import YOLO

# Load model
model = YOLO('path/to/your/yolov8-model.pt')

# Predict on a frame
results = model.predict(source='frame.jpg')

# Filter for traffic lights
traffic_lights = results.pandas().xyxy[0].query("name == 'traffic light'")

# Assuming single traffic light detection, calculate size (as proxy for distance)
if not traffic_lights.empty:
    box_area = (traffic_lights.iloc[0]['xmax'] - traffic_lights.iloc[0]['xmin']) * (traffic_lights.iloc[0]['ymax'] - traffic_lights.iloc[0]['ymin'])
    print(f"Detected traffic light box area: {box_area}")
else:
    print("No traffic light detected in frame.")
```

Please modify the `'path/to/your/yolov8-model.pt'` and `'frame.jpg'` to fit your actual file paths. Note that this method assumes the area of the detected bounding box can serve as a proxy for distance, which might require calibration based on your specific use case. 

For a more sophisticated approach especially dealing with the velocity of the car, you might consider additional sensors or data inputs that could provide velocity or displacement metrics.

Hope this gets you rolling in the right direction! Let us know if there’s anything else we can help with. 🛣️

**Reply by SMRaza1009** [reply]
Dear @pderrenger ,

Thank you so much for providing code.

**Reply by SMRaza1009** [reply]
Dear @pderrenger 

There is some problem in the code you provided.

I am getting this error in the end of code.

#My code:

from ultralytics import YOLO
import pandas


model = YOLO('yolov8n.pt')

results = model.predict(source='input_video.mp4')

traffic_lights = results.pandas().xyxy[0].query("name == 'traffic light'")

if not traffic_lights.empty:
    box_area = (traffic_lights.iloc[0]['xmax'] - traffic_lights.iloc[0]['xmin']) * (traffic_lights.iloc[0]['ymax'] - traffic_lights.iloc[0]['ymin'])
    print(f"Detected traffic light box area: {box_area}")
else:
    print("No traffic light detected in frame.")


Traceback (most recent call last):
  File "main-1.py", line 11, in <module>
    traffic_lights = results.pandas().xyxy[0].query("name == 'traffic light'")
AttributeError: 'list' object has no attribute 'pandas'

**Reply by pderrenger** [reply]
Hey there! 😊 It looks like the issue stems from `results` being a list rather than a single `Results` object. Let's tweak your code to make sure we access the correct attributes:

```python
from ultralytics import YOLO
import pandas

model = YOLO('yolov8n.pt')

# If source is a video or multiple images, results will be a list
results = model.predict(source='input_video.mp4')[0]  # Accessing the first result if it's a video or image series

traffic_lights = results.pandas().xyxy[0].query("name == 'traffic light'")

if not traffic_lights.empty:
    box_area = (traffic_lights.iloc[0]['xmax'] - traffic_lights.iloc[0]['xmin']) * (traffic_lights.iloc[0]['ymax'] - traffic_lights.iloc[0]['ymin'])
    print(f"Detected traffic light box area: {box_area}")
else:
    print("No traffic light detected in frame.")
```

If working with multiple images or a video, remember that `model.predict(source=...)` returns a list of `Results`. I've adjusted the code to extract the first set of results, assuming you're working with a video or a batch of images. Let me know if this helps or if you're encountering something different! 🚀

**Reply by Uuo10** [reply]
Hi there! I want to detect the distance from a traffic light to a car in a video, but this doesn't apply to YOLOv8, showing an AttributeError: 'Results' object has no attribute 'pandas'. See valid attributes below. How can I fix this?

**Reply by pderrenger** [reply]
Hi! Thanks for reaching out. 🙌 It sounds like you're getting an `AttributeError` because the `Results` object doesn't directly offer a `.pandas()` method as you expected. Instead, you can work with the results directly from the `Results` objects. Here's a quick way to filter for traffic lights and estimate distances by utilizing the bounding box sizes:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/your/model.pt')

# Predict on a video
results = model.predict(source='path/to/your/video.mp4')

# Iterate through results to find traffic light boxes
for result in results:
    if result.names[result.pred[:, -1].int()[0]] == 'traffic light':
        box = result.pred[0]
        box_area = (box[2] - box[0]) * (box[3] - box[1])
        print(f"Traffic light box area: {box_area}")
        # You can use box_area or box dimensions as a proxy for distance
```

Remember to adjust `'path/to/your/model.pt'` and `'path/to/your/video.mp4'` to your actual paths. This code snippet assumes you have a basic understanding of how bounding box area might relate to distance. Fine-tuning will be necessary for accurate distance estimations! Let me know if this helps or if you need further assistance. 🚀

**Reply by Uuo10** [reply]
Hi it still shows error like this: AttributeError: 'Results' object has no attribute 'pred'. See valid attributes below.

**Reply by pderrenger** [reply]
@Uuo10 hey there! 😊 My apologies for the confusion earlier. It seems there was a mistake in my previous reply. The correct way to access detection results in YOLOv8 is through the `.xyxy` attribute for bounding box coordinates. Here's an updated snippet:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/model.pt')

# Predict on an image or video
results = model.predict(source='path/to/image_or_video.jpg')

# Bounding boxes are stored in .xyxy[0], assuming the first result
for box in results.xyxy[0]:
    print(f"Detected bounding box: {box}")
    # Each 'box' contains [x1, y1, x2, y2, conf, cls]
```

Hope this clears things up! Let me know if there's anything else you need. 🚀

**Comment by CrispyHands** [comment]
> Yeah I do, do tell me your id, ill connect with you.
![image](https://github.com/ultralytics/ultralytics/assets/37016899/0ced4e54-a0c5-4bb9-9c91-2ff0ef138949)

sorry wrong ID, this the proper one.

**Reply by ahillsavio1** [reply]
Cool, I've sent you a request. (ID: TeenPro)

**Reply by pderrenger** [reply]
@ahillsavio1 i'm glad to hear you're using YOLOv8 for your custom dataset predictions! To calculate the Euclidean distance between bounding boxes, you'll first need to extract the coordinates of the boxes. Here's a concise way to do that:

```python
from ultralytics import YOLO

# Load your trained model
model = YOLO('path/to/your/model.pt')

# Perform prediction on an image
results = model('path/to/your/image.jpg')

# Extract bounding box coordinates
boxes = results.boxes.xyxy[0]  # xyxy format

# Now you can calculate the Euclidean distance between two bounding boxes
# For example, between the first and second box
import numpy as np

def euclidean_distance(box1, box2):
    center1 = np.array([(box1[0] + box1[2]) / 2, (box1[1] + box1[3]) / 2])
    center2 = np.array([(box2[0] + box2[2]) / 2, (box2[1] + box2[3]) / 2])
    distance = np.linalg.norm(center1 - center2)
    return distance

# Calculate distance between the first two boxes
if len(boxes) > 1:
    distance = euclidean_distance(boxes[0], boxes[1])
    print(f"Distance between boxes: {distance}")
```

This code snippet assumes you have at least two detected objects in your image. The `euclidean_distance` function calculates the distance between the centers of two bounding boxes. Remember to adjust the path to your model and image accordingly.

For further details on using the model and accessing the results, you can refer to the Predict mode in the Ultralytics Docs. Keep up the great work with YOLOv8! 🚀
